### DISCUSSION TITLE
Convert to LAB and apply filter to the Lightness channel

### DISCUSSION THREAD

**Discussion by Build-0-Matic** [ImageMagick/ImageMagick#5703] [Help]
Hi!
I'm writing a script and I'm having a bit of trouble.
I'm currently using Version: ImageMagick 7.1.0-43 Q16-HDRI x64 c95ef31:20220709 on Windows 10

In my script, I want to take an image in some format, change the colorspace to Lab, select just the Lightness layer and apply an unsharp mask to it before recombining everything together and re-converting to an sRGB colorspace (that will work in Photoshop).
So far, this is what I've come-up with... but it doesn't work.
`magick mogrify -format mpc -colorspace Lab -channel L -separate -unsharp %Radius%x%Sigma%+%Gain%+%Treshold% -combine "%temp%\F2D\*.mpc"
magick mogrify -format mpc -colorspace sRGB "%temp%\F2D\*.mpc"
`

The files have already been converted to .mpc/.cache when I do this because I resize my images a bit in the hope that they will go from 8 bit to 16 bit color... I haven't checked if it works.

Any help will be very welcome.

**Comment by fmw42** [comment]
You separated only the L channel and did unsharp processing. But then you try to recombine it, but only have the L channel.  You should separate all the channels, then process the L channel, then recombine.

I do not think you can use mogrify for that. You will need to write a script loop over each image separately using magick rather than magick mogrify.

The magick command can be one command in place of the two. You can combine those commands into one.

```
magick image -colorspace LAB -separate ^
( -clone 0 -unsharp %Radius%x%Sigma%+%Gain%+%Treshold% ) ^
+swap 0,3 +delete ^
-set colorspace LAB -combine -colorspace sRGB ^
result

```

You will have to do the script FOR loop over each image. But I believe my command above should work on an image by image basis.

See parenthesis processing https://imagemagick.org/Usage/basics/#parenthesis

**Reply by Build-0-Matic** [reply]
> You separated only the L channel and did unsharp processing. But then you try to recombine it, but only have the L channel. You should separate all the channels, then process the L channel, then recombine.

Yeah, that was one of the problems... though nothing really worked as it keeps looking for a file names L... or just throwing random errors.
I'll have to try this using a for /f loop.

Thanks

**Comment by snibgo** [comment]
As fmw42 says, but we can restrict the unsharp to just one channel, which is simpler and faster than separate/combine.

I suggest:
```
magick in.mpc -colorspace Lab -channel 0 -unsharp %Radius%x%Sigma%+%Gain%+%Treshold% +channel -colorspace sRGB out.mpc
```

**Reply by Build-0-Matic** [reply]
I didn't know that you could restrict the unsharp mask to a single channel. I must admit that they don't make that extra clear in the documentation...
I'll have to try this too as with this script, any gain in speed is welcome.

**Reply by Build-0-Matic** [reply]
I have been re-reading the docs for something like the third time and it seems that -unsharp can't work with the -channel function.
But I'll have to see if it really does.

**Comment by snibgo** [comment]
To demonstrate "-unsharp" on a single channel, with a massive sigma so the difference is obvious:
```
%IMG7%magick toes.png -unsharp 0x30 x0.png
%IMG7%magick toes.png -channel R -unsharp 0x30 +channel x1.png

%IMG7%magick x0.png -size 20x1 xc: x1.png +append x3.png
```
x3.png is:

![x3](https://user-images.githubusercontent.com/33812368/199131715-6b69d069-72af-4f2f-9c2d-ff38dc5b22e9.png)
Observe the difference between the outputs.

**Reply by Build-0-Matic** [reply]
It obviously works, it's just that I find it odd that it wasn't included in the docs...

**Reply by fmw42** [reply]
Included where? What file/page of the documentation should it be put? Would that have to include every function on the options page?

**Reply by Build-0-Matic** [reply]
I was checking the command line options page (https://imagemagick.org/script/command-line-options.php)
under the "-channel type" section.
It clearly specifies

> Options that are affected by the [-channel](https://imagemagick.org/script/command-line-options.php#channel) setting include the following. [-auto-gamma](https://imagemagick.org/script/command-line-options.php#auto-gamma), [-auto-level](https://imagemagick.org/script/command-line-options.php#auto-level), [-black-threshold](https://imagemagick.org/script/command-line-options.php#black-threshold), [-blur](https://imagemagick.org/script/command-line-options.php#blur), [-clamp](https://imagemagick.org/script/command-line-options.php#clamp), [-clut](https://imagemagick.org/script/command-line-options.php#clut), [-combine](https://imagemagick.org/script/command-line-options.php#combine), [-composite](https://imagemagick.org/script/command-line-options.php#composite) (Mathematical compose methods only), [-convolve](https://imagemagick.org/script/command-line-options.php#convolve), [-contrast-stretch](https://imagemagick.org/script/command-line-options.php#contrast-stretch), [-evaluate](https://imagemagick.org/script/command-line-options.php#evaluate), [-function](https://imagemagick.org/script/command-line-options.php#function), [-fx](https://imagemagick.org/script/command-line-options.php#fx), [-gaussian-blur](https://imagemagick.org/script/command-line-options.php#gaussian-blur), [-hald-clut](https://imagemagick.org/script/command-line-options.php#hald-clut), [-motion-blur](https://imagemagick.org/script/command-line-options.php#motion-blur), [-morphology](https://imagemagick.org/script/command-line-options.php#morphology), [-negate](https://imagemagick.org/script/command-line-options.php#negate), [-normalize](https://imagemagick.org/script/command-line-options.php#normalize), [-ordered-dither](https://imagemagick.org/script/command-line-options.php#ordered-dither), [-rotational-blur](https://imagemagick.org/script/command-line-options.php#rotational-blur), [-random-threshold](https://imagemagick.org/script/command-line-options.php#random-threshold), [-range-threshold](https://imagemagick.org/script/command-line-options.php#range-threshold), [-separate](https://imagemagick.org/script/command-line-options.php#separate), [-threshold](https://imagemagick.org/script/command-line-options.php#threshold), and [-white-threshold](https://imagemagick.org/script/command-line-options.php#white-threshold).

But no -unsharp...

**Reply by Build-0-Matic** [reply]
Or just mention that the -channel option can be used to apply sharpening or other filters to a single or multiple isolated channels.

**Reply by fmw42** [reply]
It says include. That does not limit it. It just is a list of some of them. I don't think there is a comprehensive list other than possibly in the code.

**Reply by Build-0-Matic** [reply]
It might be worth simply adding 

> (but not limited to)
That would definitely open the possibilities...
